File: backend/app/schemas/evaluation.py
"""评测 API 契约实现（specs/012，契约主定义 contracts/evaluation-api.md）。

字段与枚举 MUST 与契约逐条对齐，变更先改契约。
枚举唯一主定义：contracts/evaluation-api.md §enums（本文件只做消费级定义）。
"""

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

# ---- 枚举与常量（契约唯一主定义）----

EvaluatorTypeLiteral = Literal["llm_judge", "exact_match"]
EVALUATOR_TYPES: tuple[str, ...] = ("llm_judge", "exact_match")

TaskStatusLiteral = Literal["pending"]

RunStatusLiteral = Literal[
    "pending", "running", "paused", "completed", "cancelled", "failed", "interrupted",
]
RUN_STATUSES: tuple[str, ...] = (
    "pending", "running", "paused", "completed", "cancelled", "failed", "interrupted",
)

CaseRunStatusLiteral = Literal[
    "pending", "running", "passed", "failed", "execution_failed", "judge_failed", "cancelled",
]
CASE_RUN_STATUSES: tuple[str, ...] = (
    "pending", "running", "passed", "failed", "execution_failed", "judge_failed", "cancelled",
)

CaseErrorTypeLiteral = Literal[
    "agent_unavailable", "model_error", "timeout", "cancelled",
    "interrupted", "judge_invalid_output", "judge_model_error", "internal",
]

# 运行状态机合法跳转（data-model.md §4.2，实现侧唯一消费入口）
RUN_STATUS_TRANSITIONS: dict[str, tuple[str, ...]] = {
    "pending": ("running", "cancelled"),
    "running": ("paused", "completed", "cancelled", "failed", "interrupted"),
    "paused": ("running", "cancelled", "interrupted"),
    "completed": (),
    "cancelled": (),
    "failed": (),
    "interrupted": ("running",),
}

# CaseRun 可重试状态（契约 §5 retry）
RETRYABLE_CASE_RUN_STATUSES: frozenset[str] = frozenset(
    {"execution_failed", "judge_failed", "failed", "cancelled"},
)

# 评分器默认配置
DEFAULT_PASS_THRESHOLD = 80
JUDGE_DEFAULT_TEMPERATURE = "0.0"
JUDGE_DEFAULT_MAX_TOKENS = 1024

# 分页（契约 §5 GET /runs）
RUN_PAGE_DEFAULT = 1
RUN_PAGE_SIZE_DEFAULT = 20
RUN_PAGE_SIZE_MAX = 100


# ---- 数据集与用例 ----


class DatasetSaveRequest(BaseModel):
    """创建/更新数据集请求。"""

    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=500)

    @field_validator("name")
    @classmethod
    def _name_not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("数据集名称不能为空白")
        return value.strip()


class DatasetOut(BaseModel):
    """数据集输出（列表项与详情同构）。"""

    id: int
    name: str
    description: str
    case_count: int = 0
    created_at: str
    updated_at: str


class CaseSaveRequest(BaseModel):
    """创建/更新用例请求（expected_answer/scoring_criteria 可空）。"""

    user_question: str
    expected_answer: str | None = None
    scoring_criteria: str | None = None

    @field_validator("user_question")
    @classmethod
    def _question_not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("用户问题不能为空白")
        return value.strip()


class CaseOut(BaseModel):
    """用例输出。"""

    id: int
    dataset_id: int
    user_question: str
    expected_answer: str | None
    scoring_criteria: str | None
    created_at: str
    updated_at: str


class ImportResult(BaseModel):
    """导入结果（整体拒绝时 errors 携带定位信息，HTTP 422）。"""

    imported_cases: int
    skipped_cases: int = 0
    errors: list[str] = []


# ---- 评测任务 ----


class EvaluatorConfig(BaseModel):
    """评分器配置（契约 §4；exact_match 仅接受空配置）。"""

    model_config = {"extra": "allow"}

    model_model_id: int | None = None
    temperature: float | None = None
    max_tokens: int | None = None
    prompt_template: str | None = None


class TaskSaveRequest(BaseModel):
    """创建评测任务请求（契约 §4）。"""

    name: str = Field(min_length=1, max_length=100)
    agent_id: int
    dataset_id: int
    evaluator_type: EvaluatorTypeLiteral
    evaluator_config: dict[str, Any] = Field(default_factory=dict)
    pass_threshold: int = Field(default=DEFAULT_PASS_THRESHOLD, ge=0, le=100)

    @field_validator("name")
    @classmethod
    def _name_not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("任务名称不能为空白")
        return value.strip()


class TaskSummary(BaseModel):
    """任务列表项（不含三快照大字段）。"""

    id: int
    name: str
    agent_id: int
    agent_name: str
    dataset_id: int
    dataset_name: str
    evaluator_type: str
    evaluator_config: dict[str, Any]
    pass_threshold: int
    status: str
    case_count: int
    last_run_id: int | None = None
    last_run_status: str | None = None
    created_at: str
    updated_at: str


class TaskOut(TaskSummary):
    """任务详情（含三快照）。"""

    agent_snapshot: dict[str, Any]
    dataset_snapshot: dict[str, Any]
    evaluator_snapshot: dict[str, Any]


# ---- 评测运行 ----


class EvaluationRunOut(BaseModel):
    """评测运行输出（契约 §5 EvaluationRunOut）。"""

    id: int
    task_id: int
    run_id: str
    status: str
    started_at: str | None = None
    finished_at: str | None = None
    interrupted_at: str | None = None
    interrupted_reason: str | None = None
    total_cases: int = 0
    completed_cases: int = 0
    passed_cases: int = 0
    failed_cases: int = 0
    execution_failed_cases: int = 0
    judge_failed_cases: int = 0
    cancelled_cases: int = 0
    average_score: int | None = None
    pass_rate: int | None = None
    total_duration_ms: int | None = None
    total_tokens: int | None = None
    created_at: str


class RunListResponse(BaseModel):
    """GET /api/evaluation/runs 响应。"""

    items: list[EvaluationRunOut]
    total: int
    page: int
    page_size: int


class SnapshotCaseOut(BaseModel):
    """快照 Case 内容（从任务 dataset_snapshot 提取）。"""

    case_id: int
    user_question: str
    expected_answer: str | None = None
    scoring_criteria: str | None = None


class CaseRunOut(BaseModel):
    """用例运行输出（契约 §5 CaseRunOut）。"""

    id: int
    evaluation_run_id: int
    dataset_case_id: int
    status: str
    agent_run_id: str | None = None
    score: int | None = None
    reason: str | None = None
    evaluator_type: str | None = None
    evaluator_metadata: dict[str, Any] | None = None
    duration_ms: int | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    tool_call_count: int | None = None
    model_call_count: int | None = None
    iteration_count: int | None = None
    error_type: str | None = None
    error_message: str | None = None
    attempt: int = 1
    started_at: str | None = None
    finished_at: str | None = None
    case_index: int | None = None
    snapshot_case: SnapshotCaseOut | None = None


class RetryRequest(BaseModel):
    """失败 Case 重试请求（契约 §5 retry）。"""

    case_run_id: int


class RetryResponse(BaseModel):
    """重试响应。"""

    run: EvaluationRunOut
    case_run: CaseRunOut


class OkResponse(BaseModel):
    """删除/控制类操作的统一确认。"""

    ok: bool = True

File: frontend/src/api/evaluation.ts
/**
 * 评测 API 封装（specs/012，契约主定义 specs/012-agent-evaluation/contracts/evaluation-api.md）。
 * 类型 MUST 与后端 schemas/evaluation.py 逐字段对齐，变更先改契约。
 */
import { http } from './request'

// ---- 枚举与常量（契约唯一主定义，实现侧只消费）----

export type EvaluatorType = 'llm_judge' | 'exact_match'
export const EVALUATOR_TYPES: EvaluatorType[] = ['llm_judge', 'exact_match']

export type EvalRunStatus =
  | 'pending' | 'running' | 'paused'
  | 'completed' | 'cancelled' | 'failed' | 'interrupted'
export const EVAL_RUN_STATUSES: EvalRunStatus[] = [
  'pending', 'running', 'paused', 'completed', 'cancelled', 'failed', 'interrupted',
]

export type CaseRunStatus =
  | 'pending' | 'running' | 'passed' | 'failed'
  | 'execution_failed' | 'judge_failed' | 'cancelled'

export const EVAL_RUN_STATUS_META: Record<EvalRunStatus, { text: string; color: string }> = {
  pending: { text: '待运行', color: 'default' },
  running: { text: '运行中', color: 'processing' },
  paused: { text: '已暂停', color: 'warning' },
  completed: { text: '已完成', color: 'success' },
  cancelled: { text: '已取消', color: 'default' },
  failed: { text: '失败', color: 'error' },
  interrupted: { text: '已中断', color: 'warning' },
}

export const CASE_RUN_STATUS_META: Record<CaseRunStatus, { text: string; color: string }> = {
  pending: { text: '待执行', color: 'default' },
  running: { text: '执行中', color: 'processing' },
  passed: { text: '通过', color: 'success' },
  failed: { text: '未通过', color: 'error' },
  execution_failed: { text: '执行失败', color: 'error' },
  judge_failed: { text: '评分失败', color: 'warning' },
  cancelled: { text: '已取消', color: 'default' },
}

// ---- 数据结构 ----

export interface Dataset {
  id: number
  name: string
  description: string
  case_count: number
  created_at: string
  updated_at: string
}

export interface DatasetCase {
  id: number
  dataset_id: number
  user_question: string
  expected_answer: string | null
  scoring_criteria: string | null
  created_at: string
  updated_at: string
}

export interface ImportResult {
  imported_cases: number
  skipped_cases: number
  errors: string[]
}

export interface SnapshotCase {
  case_id: number
  user_question: string
  expected_answer: string | null
  scoring_criteria: string | null
}

export interface EvalTask {
  id: number
  name: string
  agent_id: number
  agent_name: string
  dataset_id: number
  dataset_name: string
  evaluator_type: string
  evaluator_config: Record<string, unknown>
  pass_threshold: number
  status: string
  case_count: number
  last_run_id: number | null
  last_run_status: string | null
  created_at: string
  updated_at: string
  // 详情专有（三快照）
  agent_snapshot?: Record<string, unknown>
  dataset_snapshot?: { cases?: SnapshotCase[] } & Record<string, unknown>
  evaluator_snapshot?: Record<string, unknown>
}

export interface EvalTaskSaveRequest {
  name: string
  agent_id: number
  dataset_id: number
  evaluator_type: EvaluatorType
  evaluator_config: { model_model_id?: number; temperature?: number; max_tokens?: number; prompt_template?: string | null }
  pass_threshold: number
}

export interface EvalRun {
  id: number
  task_id: number
  run_id: string
  status: EvalRunStatus
  started_at: string | null
  finished_at: string | null
  interrupted_at: string | null
  interrupted_reason: string | null
  total_cases: number
  completed_cases: number
  passed_cases: number
  failed_cases: number
  execution_failed_cases: number
  judge_failed_cases: number
  cancelled_cases: number
  average_score: number | null
  pass_rate: number | null
  total_duration_ms: number | null
  total_tokens: number | null
  created_at: string
}

export interface EvalRunListResponse {
  items: EvalRun[]
  total: number
  page: number
  page_size: number
}

export interface CaseRun {
  id: number
  evaluation_run_id: number
  dataset_case_id: number
  status: CaseRunStatus
  agent_run_id: string | null
  score: number | null
  reason: string | null
  evaluator_type: string | null
  evaluator_metadata: Record<string, unknown> | null
  duration_ms: number | null
  input_tokens: number | null
  output_tokens: number | null
  total_tokens: number | null
  tool_call_count: number | null
  model_call_count: number | null
  iteration_count: number | null
  error_type: string | null
  error_message: string | null
  attempt: number
  started_at: string | null
  finished_at: string | null
  case_index: number | null
  snapshot_case: SnapshotCase | null
}

// ---- 数据集 ----

export async function listDatasets(): Promise<Dataset[]> {
  return http.get<Dataset[]>('/evaluation/datasets')
}

export async function createDataset(payload: { name: string; description: string }): Promise<Dataset> {
  return http.post<Dataset>('/evaluation/datasets', payload)
}

export async function updateDataset(id: number, payload: { name: string; description: string }): Promise<Dataset> {
  return http.put<Dataset>(`/evaluation/datasets/${id}`, payload)
}

export async function deleteDataset(id: number): Promise<void> {
  await http.del(`/evaluation/datasets/${id}`)
}

export async function listDatasetCases(datasetId: number): Promise<DatasetCase[]> {
  return http.get<DatasetCase[]>(`/evaluation/datasets/${datasetId}/cases`)
}

export async function createDatasetCase(
  datasetId: number,
  payload: { user_question: string; expected_answer?: string | null; scoring_criteria?: string | null },
): Promise<DatasetCase> {
  return http.post<DatasetCase>(`/evaluation/datasets/${datasetId}/cases`, payload)
}

export async function updateDatasetCase(
  datasetId: number,
  caseId: number,
  payload: { user_question: string; expected_answer?: string | null; scoring_criteria?: string | null },
): Promise<DatasetCase> {
  return http.put<DatasetCase>(`/evaluation/datasets/${datasetId}/cases/${caseId}`, payload)
}

export async function deleteDatasetCase(datasetId: number, caseId: number): Promise<void> {
  await http.del(`/evaluation/datasets/${datasetId}/cases/${caseId}`)
}

export async function importDatasetCases(datasetId: number, payload: unknown): Promise<ImportResult> {
  return http.post<ImportResult>(`/evaluation/datasets/${datasetId}/import`, payload)
}

export function exportDatasetUrl(datasetId: number): string {
  return `/api/evaluation/datasets/${datasetId}/export`
}

// ---- 评测任务 ----

export async function listEvalTasks(): Promise<EvalTask[]> {
  return http.get<EvalTask[]>('/evaluation/tasks')
}

export async function createEvalTask(payload: EvalTaskSaveRequest): Promise<EvalTask> {
  return http.post<EvalTask>('/evaluation/tasks', payload)
}

export async function getEvalTask(id: number): Promise<EvalTask> {
  return http.get<EvalTask>(`/evaluation/tasks/${id}`)
}

export async function deleteEvalTask(id: number): Promise<void> {
  await http.del(`/evaluation/tasks/${id}`)
}

// ---- 评测运行 ----

export interface EvalRunListParams {
  task_id?: number
  status?: EvalRunStatus
  page?: number
  page_size?: number
}

export async function listEvalRuns(params: EvalRunListParams = {}): Promise<EvalRunListResponse> {
  const search = new URLSearchParams()
  if (params.task_id != null) search.set('task_id', String(params.task_id))
  if (params.status) search.set('status', params.status)
  if (params.page != null) search.set('page', String(params.page))
  if (params.page_size != null) search.set('page_size', String(params.page_size))
  const query = search.toString()
  return http.get<EvalRunListResponse>(`/evaluation/runs${query ? `?${query}` : ''}`)
}

export async function getEvalRun(id: number): Promise<EvalRun> {
  return http.get<EvalRun>(`/evaluation/runs/${id}`)
}

export async function listEvalRunCases(runId: number): Promise<CaseRun[]> {
  return http.get<CaseRun[]>(`/evaluation/runs/${runId}/cases`)
}

export async function getEvalRunCase(runId: number, caseRunId: number): Promise<CaseRun> {
  return http.get<CaseRun>(`/evaluation/runs/${runId}/cases/${caseRunId}`)
}

export async function startEvalRun(taskId: number): Promise<EvalRun> {
  return http.post<EvalRun>(`/evaluation/tasks/${taskId}/runs`)
}

export async function pauseEvalRun(runId: number): Promise<EvalRun> {
  return http.post<EvalRun>(`/evaluation/runs/${runId}/pause`)
}

export async function resumeEvalRun(runId: number): Promise<EvalRun> {
  return http.post<EvalRun>(`/evaluation/runs/${runId}/resume`)
}

export async function cancelEvalRun(runId: number): Promise<EvalRun> {
  return http.post<EvalRun>(`/evaluation/runs/${runId}/cancel`)
}

export async function retryEvalRunCase(runId: number, caseRunId: number): Promise<{ run: EvalRun; case_run: CaseRun }> {
  return http.post(`/evaluation/runs/${runId}/retry`, { case_run_id: caseRunId })
}
